### PR TITLE
Support Pre-Hook Enabled ERC-1271 Signature Verification

### LINKED ISSUE #1599: Consider Pre-Interactions for EIP-1271 Signature Verification
Currently EIP-1271 signature verification done for order creation and auction cutting assumes that the signature is valid given the current state of the blockchain. However, it is possible for pre-interactions to do the required setup in order for the EIP-1271 signature to be valid. As such, we should execute pre-interactions in the simulation that we do for verifying EIP-1271 signatures.

### PR DESCRIPTION
**Pull request by nlordell**
Fixes #1599.

This PR adds new simulation logic for verifying ERC-1271 signatures in a way that considers pre-hooks. This opens up the possibility of gas-less smart contract orders (where the smart contract order setup is done as part of a pre-hook).

Currently, this code isn't used anywhere. In a follow up PR I will add the configuration knobs required to enable this signature validation and perform a manual E2E test there.

### Test Plan

Automated testing blocked on https://github.com/cowprotocol/services/issues/1826. Will do a manual test in the aforementioned follow-up PR.

### CODE REVIEW

File: crates/contracts/solidity/interfaces/IERC20.sol [hunk 1]
```diff
@@ -11,11 +11,6 @@ interface IERC20 {
     function transferFrom(address, address, uint256) external returns (bool);
 }
 
-/// @title Phony token interface
-interface IPhonyERC20 is IERC20 {
-    function mintPhonyTokens(address, uint256) external returns (bool);
-}
-
```
Comment (nlordell) [inline crates/contracts/solidity/interfaces/IERC20.sol:18]: Unrelated cleanup. Just noticed when browsing our Solidity code.

File: crates/contracts/solidity/Balances.sol [hunk 2]
```diff
@@ -8,7 +8,7 @@ import { Transfer, IVaultRelayer } from "./interfaces/IVaultRelayer.sol";
 
 /// @title A contract for simulating available balances for settlements.
 contract Balances {
-    /// The on-chain 
+    /// @dev The on-chain contracts required for the simulation.
```
Comment (nlordell) [inline crates/contracts/solidity/Balances.sol:11]: Unrelated comment fix.

File: crates/shared/src/signature_validator/simulation.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    async fn simulate(
+        &self,
+        check: &SignatureCheck,
+    ) -> Result<Verification, SignatureValidationError> {
+        // We simulate the signature verification from the Settlement contract's
+        // context. This allows us to check:
+        // 1. How the pre-interactions would behave as part of the settlement
+        // 2. Simulate the actual `isValidSignature` calls that would happen as part of
+        //    a settlement
+        let signatures = dummy_contract!(contracts::support::Signatures, self.settlement);
```
Comment (fleupold) [inline crates/shared/src/signature_validator/simulation.rs:43]: I remember we made the `SettlementContract` inherit from `StorageAccessible` which should allow these types of simulations without requiring any fancy simulator/storage override if I'm not mistaken (using [`simulateDelegateCall`](https://github.com/cowprotocol/contracts/blob/main/src/contracts/mixins/StorageAccessible.sol#L60C14-L60C34)). 
Is there a particular reason we are not using this with a standard eth_call?

Would this also allow us to do TradeVerification on Gnosis Chain without a special node?

**Review by fleupold — APPROVED** [review]
Nice and slick component. Just one question if it would make sense to do this with standard RPC calls?

File: crates/shared/src/signature_validator/simulation.rs (hunk 3 above)
Comment (nlordell) [reply crates/shared/src/signature_validator/simulation.rs:43]: I had completely forgotten about this! Yes, we can change the `Balances` and `Signatures` contracts to work this way.

It won't work with `TradeVerification` however, because that requires impersonating the trader and the solver, and not just the settlement contract.

**Comment by nlordell** [comment]
> Just one question if it would make sense to do this with standard RPC calls?

It would - and great idea! We can even do it without deploying contracts on-chain with "constructor call magic". PR incoming...

File: crates/shared/src/signature_validator/simulation.rs (hunk 3 above)
Comment (nlordell) [reply crates/shared/src/signature_validator/simulation.rs:43]: Will do this in a follow-up.

**Review by MartinquaXD — APPROVED** [review]
Approving since you already said you wanted to redo the logic with `StorageAccessible`.
